src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,469 @@
+
+                    Vector128<float> result = Sse.Add(srcVector, dstVector);
+                    Sse.Store(pDstCurrent, result);
+
+                    pSrcCurrent += 4;
+                    pDstCurrent += 4;
+                }
+
+                while (pSrcCurrent < pEnd)
+                {
+                    Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                    Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                    Vector128<float> result = Sse.AddScalar(srcVector, dstVector);
+                    Sse.StoreScalar(pDstCurrent, result);
+
+                    pSrcCurrent++;
+                    pDstCurrent++;
+                }
+            }
+        }
+
+        internal static unsafe void AddSU(Span<float> src, Span<int> idx, Span<float> dst)
+        {
+            fixed (float* psrc = src)
+            fixed (int* pidx = idx)
+            fixed (float* pdst = dst)
+            {
+                float* pSrcCurrent = psrc;
+                int* pIdxCurrent = pidx;
+                float* pDstCurrent = pdst;
+                int* pEnd = pidx + idx.Length;
+
+                while (pIdxCurrent + 4 <= pEnd)
+                {
+                    Vector128<float> srcVector = Load4(pDstCurrent, pIdxCurrent);
+                    Vector128<float> dstVector = Sse.LoadVector128(pSrcCurrent);
+
+                    srcVector = Sse.Add(srcVector, dstVector);
+                    Store4(srcVector, pDstCurrent, pIdxCurrent);
+
+                    pIdxCurrent += 4;
+                    pSrcCurrent += 4;
+                }
+
+                while (pIdxCurrent < pEnd)
+                {
+                    pDstCurrent[*pIdxCurrent] += *pSrcCurrent;
+
+                    pIdxCurrent++;
+                    pSrcCurrent++;
+                }
+            }
+        }
+
+        internal static unsafe void MulElementWiseU(Span<float> src1, Span<float> src2, Span<float> dst)
+        {
+            fixed (float* psrc1 = &src1[0])
+            fixed (float* psrc2 = &src2[0])
+            fixed (float* pdst = dst)
+            {
+                float* pSrc1Current = psrc1;
+                float* pSrc2Current = psrc2;
+                float* pDstCurrent = pdst;
+                float* pEnd = pdst + dst.Length;
+
+                while (pDstCurrent + 4 <= pEnd)
+                {
+                    Vector128<float> src1Vector = Sse.LoadVector128(pSrc1Current);
+                    Vector128<float> src2Vector = Sse.LoadVector128(pSrc2Current);
+                    src2Vector = Sse.Multiply(src1Vector, src2Vector);
+                    Sse.Store(pDstCurrent, src2Vector);
+
+                    pSrc1Current += 4;
+                    pSrc2Current += 4;
+                    pDstCurrent += 4;
+                }
+
+                while (pDstCurrent < pEnd)
+                {
+                    Vector128<float> src1Vector = Sse.LoadScalarVector128(pSrc1Current);
```

Comment:
Why do any vector operations here at all?  Why not just simply do the multiplication like normal?

`*pDstCurrent = (*pSrc1Current) * (*pSrc2Current)`

Reply:
Oh, is it because that's what the existing native C++ code is doing?

---
In reply to: [206389105](https://github.com/dotnet/machinelearning/pull/562#discussion_r206389105) [](ancestors = 206389105)

Reply:
Unfortunately yes :( Link: https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L1830-L1849

Reply:
OK, that's fine.